tensorflow_model_optimization/python/core/sparsity_tf2/riglpruner.py

```diff
@@ -0,0 +1,366 @@
+      )
+    return new_connections
+
+  def _get_grow_tensor(self, weight, method, step=0):
+    if method == 'zeros':
+      grow_tensor = tf.zeros_like(weight, dtype=weight.dtype)
+    elif method == 'random_normal':
+      divisor = 1.
+      stdev = tf.math.reduce_std(weight)
+      grow_tensor = self._random_normal(step, weight.get_shape(), stddev=stdev, dtype=weight.dtype, seed=self._seed) / divisor
+    elif method == 'random_uniform':
+      mean = tf.math.reduce_mean(tf.math.abs(weight))
+      divisor = 1.
+      grow_tensor = self._random_uniform(step, weight.get_shape(), minval=-mean, maxval=mean, 
+                                        dtype=weight.dtype, seed=self._seed) / divisor
+    return grow_tensor
+
+  def _generic_top_k(self, scores, mask, n_to_modify, n_total):
+     # sort the entire array since TPU requires k to be constant
+     # so that its aggregation methods can be reliable
+    _, sorted_idx = tf.math.top_k(
+      tf.reshape(scores, [-1]), k=n_total
+    )
+    expanded_sorted_idx = tf.expand_dims(sorted_idx, 1)
+    new_values = tf.where(
+      tf.range(n_total) < n_to_modify,
+      tf.ones_like(sorted_idx, dtype=mask.dtype),
+      tf.zeros_like(sorted_idx, dtype=mask.dtype)
+    )
+    updated_mask = tf.scatter_nd(expanded_sorted_idx, new_values, new_values.shape)
+
+    return updated_mask
+
+  def _update_mask(self, step, update_fraction, mask, weight, grad):
+    """Called by _maybe_update_block_mask.
+    Updates mask based on weight and grad information.
+
+    Args:
+      step: current iteration from optimizer
+      update_fraction: the fraction of existing connections to update
+      mask: the mask to update
+      weight: trainable variable representing weights
+      grad: gradients obtained from the optimizer
+    """
+    # compute the top k magnitudes then update the current mask
+    drop_scores = self._get_drop_weights(mask, weight, noise_std=self._noise_std, step=step)
+    # need access to exactly which entries are growing to zero out optimizer slot
+    grow_scores = self._get_grow_grads(mask, grad)
+    n_total = tf.size(drop_scores)
+    n_ones = tf.cast(tf.reduce_sum(mask), dtype=tf.int32) # floor not ceiling like sparsity
+    n_prune = tf.cast(
+      tf.cast(n_ones, dtype=tf.float32) * update_fraction, tf.int32
+    )
+    n_keep = n_ones - n_prune
+
+    dropped_mask = self._generic_top_k(drop_scores, mask, n_keep, n_total)
+
+    if grow_scores is not None:
+      # flatten the scores
+      grow_scores = tf.reshape(grow_scores, (-1,))
+      # set enabled connections (ones) to min(scores) - 1, i.e. they have the lowest scores
+      grow_scores_lifted = tf.where(
+        tf.math.equal(dropped_mask, 1),
+        tf.ones_like(dropped_mask) * (tf.reduce_min(grow_scores) - 1), grow_scores
+      )
+      grown_mask = self._generic_top_k(grow_scores_lifted, mask, n_prune, n_total)
+      # ensure that masks are disjoint
+      tf.debugging.Assert(
```

Comment:
This isn't a helpful assert for the following reasons.

1. If it triggers, the error raised would be incomprehensible to users because there's no message
2. It's realistically not something we expect users to be able to control. If it happens then it's a failure in our code. This means it's better to verify this w/ our actual unit tests.

So, I'd suggest getting rid of it and making sure you have test coverage for it.